src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java

```diff
@@ -667,6 +697,9 @@ private static boolean isLoopAst(int ast) {
         /** Contains definitions of uninitialized variables. */
         private final Deque<DetailAST> uninitializedVariables = new ArrayDeque<>();
 
+        /** Whether there is a {@code break} or {@code return} in the scope. */
+        private boolean containsBreakOrReturn;
```

Comment:
@rnveach 
The first example is handled by the check just like the one I wrote at https://github.com/checkstyle/checkstyle/pull/4115#discussion_r108645070:
```
int a; // violation
while (...) {
  if (...) {
    a = 1;
    System.out.print(a);
    break;
  }
}
// never assign or access 'a' again
```
> And the check won't raise that violation by now. In fact we have a judgement here https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java#L511-L513
> > if the variable is declared outside the loop and initialized inside
the loop, then it cannot be declared final, as it can be initialized
more than once in this case

> Therefore, whether I handle break in loop or not, no violation would be raised due to that judgement.

This is a complicated problem, due to the code about handling loop I mentioned. I will get back to this later and I am willing to focus on the second example now.

For the second one, it is indeed a false negative. You are correct, `return` and `break` should be handled differently for case like that.
I am trying to fix this and find another token we ignored before: `throw`.
Example:
```
void foo3() {
  int a; // violation
  int b; // no violation
  final int random = (int) Math.random();
  switch (random) {
    case 0:
      b = 0;
    case 1:
      b = 1;
      a = 1;
      throw new RuntimeException();
    default:
      a = 2;
      break;
  } 
}
```
`throw` is smth like return if it is not surrounded with `try`.
Also, we didn't handle the fall-through of `case` before. There could be a false positive of `b`.

I am afraid we would find many many bugs of this check during this PR. 😅  It is no longer just a "return" and "break" problem.